# Proyecto 4. Exploración y limpiza de datos usando `Pandas` y `Numpy`
## "*Software para Data Science - MDS - USS*"
### Integrantes: "Camila Rodriguez O. - Paula Vargas Q. - Miguel Chandía C."

## 1. Introduccción

En este proyecto usaremos con un conjunto de datos de autos usados de **eBay** *Kleinanzeigen*, una sección de clasificados del sitio web alemán de **eBay**.

Las columnas:

-   `dateCrawled`: cuándo se rastreó este anuncio por primera vez. Todos los valores de campo se toman a partir de esta fecha.
-   `name`: nombre del automóvil.
-   `seller`: si el vendedor es privado o un comerciante.
-   `offerType`: el tipo de listado.
-   `price`: el precio del anuncio para vender el automóvil.
-   `abtest`: si la lista está incluida en una prueba A/B.
-   `vehicleType`: el tipo de vehículo.
-   `yearOfRegistration`: el año en el que se registró por primera vez el automóvil.
-   `gearbox`: el tipo de transmisión.
-   `powerPS`: la potencia del auto en PS.
-   `model`: el nombre del modelo de automóvil.
-   `kilometer`: cuántos kilómetros ha recorrido el automóvil.
-   `monthOfRegistration`: el mes en el que se registró el automóvil por primera vez.
-   `fuelType`: qué tipo de combustible usa el automóvil.
-   `brand`: la marca del automóvil.
-   `notRepairedDamage`: si el automóvil tiene un daño que aún no se ha reparado.
-   `dateCreated`: la fecha en la que se creó el anuncio de eBay.
-   `nrOfPictures`: el número de imágenes del anuncio.
-   `postalCode`: el código postal para la ubicación del vehículo.
-   `lastSeenOnline`: cuándo el rastreador vio este anuncio por última vez en línea.

Comencemos importando las bibliotecas que necesitamos y leyendo el conjunto de datos en `Pandas`.

## Instrucciones

1.  Importar las bibliotecas `Pandas` y `NumPy`.

    -   Lea el archivo `autos.csv`en `Pandas` y asígnelo al nombre de variable `autos`.
    -   Pruebe sin especificar ninguna codificación (que será `UTF-8` por defecto)
    -   Si obtiene un error de codificación, pruebe las siguientes dos codificaciones más populares (`Latin-1` y `Windows-1252`) hasta que pueda leer el archivo sin errores.

In [57]:
import pandas as pd
import numpy as np

In [58]:
csv_file = 'autos.csv'
# Lectura sin codificación
#autos = pd.read_csv(csv_file)

# Codificación Latin-1 funcionó correcta
#autos = pd.read_csv(csv_file, encoding='Latin-1')

# Codificación Windows-1252 funcionó correcta
autos = pd.read_csv(csv_file, encoding='Windows-1252')

#Codificación utf-8 no funcionó arrojó UnicodeDecodeError
#autos = pd.read_csv(csv_file,sep=',',encoding='utf8')
autos

dateCrawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  yearOfRegistration  \
0       privat   Angebot    480     test         NaN                1993   
1       privat   Angebot  18300     test       coupe                2011   
2       privat   Angebot   9800     test         suv                2004   
3       privat   Angebot   1500     test  kleinwagen                2001   
4       privat   Angebot   3600     test  kleinwagen                2008   
...        ...       ...    ...      ...         ...                 ...   
371523  privat   Angebot   2200     test         NaN                2005   
371524  privat   Angebot   1199     test      cabrio                2000   
371525  privat   Angebot   9200     test         bus                1996   
371526  privat   Angebot   3400     test       kombi                2002   
371527  privat   Angebot  28990  control   limousine                2013   

          gearbox  powerPS        model  kilometer  monthOfRegistration  \
0         manuell        0         golf     150000                    0   
1         manuell      190          NaN     125000                    5   
2       automatik      163        grand     125000                    8   
3         manuell       75         golf     150000                    6   
4         manuell       69        fabia      90000                    7   
...           ...      ...          ...        ...                  ...   
371523        NaN        0          NaN      20000                    1   
371524  automatik      101       fortwo     125000                    3   
371525    manuell      102  transporter     150000                    3   
371526    manuell      100         golf     150000                    6   
371527    manuell      320      m_reihe      50000                    8   

       fuelType           brand notRepairedDamage          dateCreated  \
0        benzin      volkswagen               NaN  2016-03-24 00:00:00   
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
0                  0       70435  2016-04-07 03:16:57  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
...              ...         ...                  ..

In [59]:
# ahora crearemos una función utilizando chardet para detectar la codificación del archivo csv

import chardet

def return_enc(raw_data):
    raw_data=open(raw_data,"rb").read()
    return chardet.detect(raw_data)

return_enc(csv_file)

{'encoding': 'Windows-1252', 'confidence': 0.73, 'language': ''}

2.  Utilice los métodos `DataFrame.info()` y `DataFrame.head()` para imprimir información sobre el dataframe `autos`, así como las primeras filas.

In [60]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

In [61]:
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

3.  Describa brevemente sus observaciones.

## 2. Limpieza de nombres de columnas

De los resultados anteeriores, podemos hacer las siguientes observaciones:

-   El conjunto de datos contiene 20 columnas, la mayoría de las cuales son cadenas de texto.
-   Algunas columnas tienen valores nulos, pero ninguna tiene más de \~20% de valores nulos.
-   Los nombres de las columnas usan *camelcase* en lugar del estilo *snakecase*, preferido de `Python`, lo que significa que no podemos simplemente reemplazar los espacios con guiones bajos.

Convierta los nombres de las columnas de *camelcase* a *snakecase* y reformule algunos de los nombres de las columnas en función del diccionario de datos para que sean más descriptivos.

## Instrucciones

1.  Utilice el atributo `DataFrame.columns` para imprimir una matriz de los nombres de columna existentes.

In [62]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

2.  Copie esa matriz y realice las siguientes modificaciones en los nombres de las columnas:

    -   `yearOfRegistration` a `registration_year`
    -   `monthOfRegistration` a `registration_month`
    -   `notRepairedDamage` a `unrepaired_damage`
    -   `dateCreated` a `ad_created`
    -   El resto de los nombres de las columnas, desde *camelcase* a *snakecase*.

In [63]:
columns_aux = autos.columns
columns_aux = []
for c in autos.columns:
    if c == 'yearOfRegistration':
        columns_aux.append('registration_year')
    elif c == 'monthOfRegistration':
        columns_aux.append('registration_month')
    elif c == 'notRepairedDamage':
        columns_aux.append('unrepaired_damage')
    else:
        columns_aux.append(c)
columns_aux

['dateCrawled',
 'name',
 'seller',
 'offerType',
 'price',
 'abtest',
 'vehicleType',
 'registration_year',
 'gearbox',
 'powerPS',
 'model',
 'kilometer',
 'registration_month',
 'fuelType',
 'brand',
 'unrepaired_damage',
 'dateCreated',
 'nrOfPictures',
 'postalCode',
 'lastSeen']

3.  Asigne los nombres de las columnas modificadas al atributo `DataFrame.columns`.

In [64]:
autos.rename(columns={'yearOfRegistration': 'registration_year','monthOfRegistration': 'registration_month','notRepairedDamage': 'unrepaired_damage','dateCreated': 'ad_created'}, inplace=True)
autos

dateCrawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  registration_year  \
0       privat   Angebot    480     test         NaN               1993   
1       privat   Angebot  18300     test       coupe               2011   
2       privat   Angebot   9800     test         suv               2004   
3       privat   Angebot   1500     test  kleinwagen               2001   
4       privat   Angebot   3600     test  kleinwagen               2008   
...        ...       ...    ...      ...         ...                ...   
371523  privat   Angebot   2200     test         NaN               2005   
371524  privat   Angebot   1199     test      cabrio               2000   
371525  privat   Angebot   9200     test         bus               1996   
371526  privat   Angebot   3400     test       kombi               2002   
371527  privat   Angebot  28990  control   limousine               2013   

          gearbox  powerPS        model  kilometer  registration_month  \
0         manuell        0         golf     150000                   0   
1         manuell      190          NaN     125000                   5   
2       automatik      163        grand     125000                   8   
3         manuell       75         golf     150000                   6   
4         manuell       69        fabia      90000                   7   
...           ...      ...          ...        ...                 ...   
371523        NaN        0          NaN      20000                   1   
371524  automatik      101       fortwo     125000                   3   
371525    manuell      102  transporter     150000                   3   
371526    manuell      100         golf     150000                   6   
371527    manuell      320      m_reihe      50000                   8   

       fuelType           brand unrepaired_damage           ad_created  \
0        benzin      volkswagen               NaN  2016-03-24 00:00:00   
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
0                  0       70435  2016-04-07 03:16:57  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
...              ...         ...                  ...  
371523             0

4.  Use `DataFrame.head()` para ver el estado actual del dataframe autos.

In [65]:
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  registration_year    gearbox  powerPS  model  \
0    480   test         NaN               1993    manuell        0   golf   
1  18300   test       coupe               2011    manuell      190    NaN   
2   9800   test         suv               2004  automatik      163  grand   
3   1500   test  kleinwagen               2001    manuell       75   golf   
4   3600   test  kleinwagen               2008    manuell       69  fabia   

   kilometer  registration_month fuelType       brand unrepaired_damage  \
0     150000                   0   benzin  volkswagen               NaN   
1     125000                   5   diesel        audi                ja   
2     125000                   8   diesel        jeep               NaN   
3     150000                   6   benzin  volkswagen              nein   
4      90000                   7   diesel       skoda              nein   

            ad_created  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

5.  Describa y explique los cambios que hizo y por qué los hizo.

## 3. Exploración y limpieza inicial.

Ahora hagamos una exploración básica de datos para determinar qué otras tareas de limpieza deben realizarse. Inicialmente buscaremos:

-   Columnas de texto donde todos o casi todos los valores son iguales. Estos a menudo se pueden descartar ya que no tienen información útil para el análisis.
-   Ejemplos de datos numéricos almacenados como texto que se pueden limpiar y convertir.

Los siguientes métodos son útiles para explorar los datos:

-   `DataFrame.describe()` (con `include = 'all'` para obtener columnas categóricas y numéricas)
-   `Series.value_counts()` y `Series.head()` si alguna columna necesita una revisión más detallada.

## Instrucciones

1.  Utilice `DataFrame.describe()` para ver las estadísticas descriptivas de todas las columnas.

In [66]:
autos.describe()

price  registration_year        powerPS      kilometer  \
count  3.715280e+05      371528.000000  371528.000000  371528.000000   
mean   1.729514e+04        2004.577997     115.549477  125618.688228   
std    3.587954e+06          92.866598     192.139578   40112.337051   
min    0.000000e+00        1000.000000       0.000000    5000.000000   
25%    1.150000e+03        1999.000000      70.000000  125000.000000   
50%    2.950000e+03        2003.000000     105.000000  150000.000000   
75%    7.200000e+03        2008.000000     150.000000  150000.000000   
max    2.147484e+09        9999.000000   20000.000000  150000.000000   

       registration_month  nrOfPictures    postalCode  
count       371528.000000      371528.0  371528.00000  
mean             5.734445           0.0   50820.66764  
std              3.712412           0.0   25799.08247  
min              0.000000           0.0    1067.00000  
25%              3.000000           0.0   30459.00000  
50%              6.000000           0.0   49610.00000  
75%              9.000000           0.0   71546.00000  
max             12.000000           0.0   99998.00000

2.  Escriba señalando:

    -   Cualquier columna que tenga principalmente un valor que sea candidato a descartarse
    -   Cualquier columna que necesite más investigación.
    -   Cualquier ejemplo de datos numéricos almacenados como texto que deba limpiarse.

3.  Si necesita investigar más alguna columna, hágalo y anote cualquier cosa adicional que encuentre.

4.  Probablemente descubrió que las columnas de precio y odómetro son valores numéricos almacenados como texto. Para cada columna:

    -   Elimine cualquier carácter no numérico.
    -   Convierta la columna a tipo numérico.
    -   Utilice `DataFrame.rename()` para cambiar el nombre de la columna a `odometer_km`.

Las preguntas 2,3 y 4 ya están desarrolladas, por que el set de datos venía con las transformaciones solicitadas d eacuerdo con respuesta del profesor.

In [67]:
autos.rename(columns={'kilometer': 'odometer_km'}, inplace=True)
autos

dateCrawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  registration_year  \
0       privat   Angebot    480     test         NaN               1993   
1       privat   Angebot  18300     test       coupe               2011   
2       privat   Angebot   9800     test         suv               2004   
3       privat   Angebot   1500     test  kleinwagen               2001   
4       privat   Angebot   3600     test  kleinwagen               2008   
...        ...       ...    ...      ...         ...                ...   
371523  privat   Angebot   2200     test         NaN               2005   
371524  privat   Angebot   1199     test      cabrio               2000   
371525  privat   Angebot   9200     test         bus               1996   
371526  privat   Angebot   3400     test       kombi               2002   
371527  privat   Angebot  28990  control   limousine               2013   

          gearbox  powerPS        model  odometer_km  registration_month  \
0         manuell        0         golf       150000                   0   
1         manuell      190          NaN       125000                   5   
2       automatik      163        grand       125000                   8   
3         manuell       75         golf       150000                   6   
4         manuell       69        fabia        90000                   7   
...           ...      ...          ...          ...                 ...   
371523        NaN        0          NaN        20000                   1   
371524  automatik      101       fortwo       125000                   3   
371525    manuell      102  transporter       150000                   3   
371526    manuell      100         golf       150000                   6   
371527    manuell      320      m_reihe        50000                   8   

       fuelType           brand unrepaired_damage           ad_created  \
0        benzin      volkswagen               NaN  2016-03-24 00:00:00   
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
0                  0       70435  2016-04-07 03:16:57  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
...              ...         ...                  ..

## 4. Exploración de las columnas odómetro y precio

En el paso anterior, encontró de que hay varias columnas de texto en las que casi todos los valores son iguales (`seller` y `offer_type`). También convirtió las columnas `price` y `odometer` a tipos numéricos y cambió el nombre de odometer a `odometer_km`.

Continuemos explorando los datos, específicamente buscando datos que no se vean bien. Comenzaremos analizando las columnas `odometer_km` y `price`. Estos son los pasos que seguiremos:

-   Analice las columnas utilizando valores mínimos y máximos y busque cualquier valor que parezca demasiado alto o bajo (valores atípicos) que podríamos querer eliminar.

-   Usaremos:

    -   `Series.unique()`. Forma para ver cuántos valores únicos.

    -   `Series.describe()` para ver mínimo/máximo/mediana/media, etc.

    -   `Series.value_counts()`, con algunas variaciones:

        -   encadenado a `.head()` si hay muchos valores.
        -   Debido a que `Series.value_counts()` devuelve una serie, podemos usa `Series.sort_index()` con `ascennding = True` o `False` para ver los valores más altos y más bajos con sus conteos (también puede encadenar a `head()`).

    -   Para eliminar valores atípicos, puede utilizar `df[(df["col"]> x) & (df["col"] < y)]`, pero es más legible usar `df[df["col"].between(x, y)]`.

## Instrucciones

1.  Para cada una de las columnas `odometer_km` y `price`:

    -   Utilice las técnicas anteriores para explorar los datos.
    -   Si encuentra que hay valores atípicos, elimínelos y escriba un párrafo explicando su decisión.
    -   Una vez que haya eliminado los valores atípicos, haga algunas observaciones sobre los valores restantes.

In [68]:
autos['odometer_km'].unique()

array([150000, 125000,  90000,  40000,  30000,  70000,   5000, 100000,
        60000,  20000,  80000,  50000,  10000])

In [69]:
autos['price'].unique()

array([  480, 18300,  9800, ..., 18429, 24895, 10985])

#### Observamos que todos los valores son positivos por tanto usaremos numpy para reducir su espacio en memoria.

In [152]:
autos.odometer_km = autos.odometer_km.astype(np.uint32)
autos.price = autos.price.astype(np.uint32)
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   dateCrawled         371528 non-null  object
 1   name                371528 non-null  object
 2   seller              371528 non-null  object
 3   offerType           371528 non-null  object
 4   price               371528 non-null  uint32
 5   abtest              371528 non-null  object
 6   vehicleType         333659 non-null  object
 7   registration_year   371528 non-null  int64 
 8   gearbox             351319 non-null  object
 9   powerPS             371528 non-null  int64 
 10  model               351044 non-null  object
 11  odometer_km         371528 non-null  uint32
 12  registration_month  371528 non-null  int64 
 13  fuelType            338142 non-null  object
 14  brand               371528 non-null  object
 15  unrepaired_damage   299468 non-null  object
 16  ad

In [153]:
autos[["odometer_km", "price"]].describe()

odometer_km         price
count  371528.000000  3.715280e+05
mean   125618.688228  1.729514e+04
std     40112.337051  3.587954e+06
min      5000.000000  0.000000e+00
25%    125000.000000  1.150000e+03
50%    150000.000000  2.950000e+03
75%    150000.000000  7.200000e+03
max    150000.000000  2.147484e+09

#### Observamos un ahorro de memoria `2,8 MB` cuando usamos el `info()` anterior `memory usage: 56.7+ MB` y ahora `memory usage: 53.9+ MB`

In [80]:
autos['odometer_km'].value_counts()

150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: odometer_km, dtype: int64

In [78]:
autos['price'].value_counts().head()

0       10778
500      5670
1500     5394
1000     4649
1200     4594
Name: price, dtype: int64

In [83]:
autos['odometer_km'].value_counts().sort_index().head()

5000     7069
10000    1949
20000    5676
30000    6041
40000    6376
Name: odometer_km, dtype: int64

In [84]:
autos['price'].value_counts().sort_index().head()

0    10778
1     1189
2       12
3        8
4        1
Name: price, dtype: int64

In [85]:
autos['price'].sort_values(ascending=False)

129969    2147483647
127820      99999999
77520       99999999
150909      99999999
56973       99999999
             ...    
125755             0
74818              0
301014             0
196452             0
310220             0
Name: price, Length: 371528, dtype: uint32

#### De los ejercicios anteriores observamos valores atipicos como 0 o 1 euros (asumimos que son euros ya que son ventas en Alemania), por lo que nos basaremos en la referencia de https://www.mexicanosenalemania.de/2016/11/23/cuanto-cuesta-tener-coche-en-alemania/ para establecer que solo considerarémos valores sobre 1.000 eur

In [155]:
autos[autos['price']>=200000]

dateCrawled  \
1846    2016-03-10 22:47:05   
10649   2016-04-05 09:36:18   
14663   2016-04-03 06:03:56   
16889   2016-03-15 16:45:45   
20143   2016-03-18 00:55:53   
...                     ...   
365287  2016-03-08 13:58:07   
365461  2016-03-31 18:53:48   
366653  2016-03-20 10:59:52   
366861  2016-03-21 10:53:06   
367041  2016-03-24 19:00:03   

                                                     name  seller offerType  \
1846    BMW_M1_Museumsfahrzeug_Neuwagenzustand_Glossy_...  privat   Angebot   
10649                       Porsche_996_GT2_MK2_Clubsport  privat   Angebot   
14663                          Tausch_moeglich_Polo_vw_n9  privat   Angebot   
16889                    Ford_Mondeo_Kombi_TÜV_abgelaufen  privat   Angebot   
20143                          Porsche_911R_Einer_von_911  privat   Angebot   
...                                                   ...     ...       ...   
365287               Porsche_991_GT3RS_Sofort_verfuegbar.  privat   Angebot   
365461   BMW_M1_mit_4900_Km_Neuwagenzustand_Glossy_Orange  privat   Angebot   
366653      BMW_E36_Cabrio_tausch_gegen_ein_Opel_Insignia  privat   Angebot   
366861         Ferrari_FXX__unfallfrei__perfekter_Zustand  privat   Angebot   
367041                                      Maserati_MC12  privat   Angebot   

           price   abtest vehicleType  registration_year    gearbox  powerPS  \
1846      579000  control       coupe               1980    manuell      277   
10649     420000  control       coupe               2004    manuell      483   
14663   11111111  control       coupe               2003    manuell       64   
16889    1000000  control       kombi               1998        NaN        0   
20143    1250000     test       coupe               2016    manuell      500   
...          ...      ...         ...                ...        ...      ...   
365287    250000  control       coupe               2016  automatik      500   
365461    599000  control       coupe               1980    manuell      377   
366653  99999999  control      cabrio               1996    manuell      192   
366861   3895000     test       coupe               2006        NaN      799   
367041   1600000     test       coupe               2005        NaN      625   

         model  odometer_km  registration_month fuelType           brand  \
1846    andere        20000                  12   benzin             bmw   
10649      911        50000                   4   benzin         porsche   
14663     polo       150000                   2   benzin      volkswagen   
16889   mondeo       150000                   0   benzin            ford   
20143      911         5000                   3   benzin         porsche   
...        ...          ...                 ...      ...             ...   
365287     911         5000                   2   benzin         porsche   
365461  andere         5000                   3   benzin             bmw   
366653     3er       150000                   0      NaN             bmw   
366861     NaN         5000                   4   benzin  sonstige_autos   
367041     NaN        10000                   4   benzin  sonstige_autos   

       unrepaired_damage           ad_created  nrOfPictures  postalCode  \
1846                nein  2016-03-10 00:00:00             0       60435   
10649               nein  2016-04-05 00:00:00             0       81669   
14663                NaN  2016-04-03 00:00:00             0       46535   
16889                 ja  2016-03-15 00:00:00             0       26607   
20143               nein  2016-03-17 00:00:00             0       70435   
...                  ...                  ...           ...         ...   
365287              nein  2016-03-08 00:00:00             0       85551   
365461              nein  2016-03-31 00:00:00             0       60435   
366653               NaN  2016-03-20 00:00:00             0       36251   
366861              nein  2016-03-21 00:00:00             0       6032

In [159]:
autos[autos['price']>=200000]['price'].sort_values(ascending=False)

129969    2147483647
108987      99999999
139052      99999999
77520       99999999
150909      99999999
             ...    
351528        208250
143805        205000
318065        205000
161652        200000
187095        200000
Name: price, Length: 172, dtype: uint32

#### Analizando los modelos sobre los 200.000 euros y buscando algúnos valores de referencia respecto algúnos modelos, por ejemplo, BMW M1 Museumsfahrzeug cotiza al rededor de los 450.000 euros o el porsche 911 GT3 al rededor de los 224.000 euros, por tanto sobre 450.000 euros será considerado sobrevaluado y podría sesgar nuestro análisis.

In [162]:
autos_valuate = autos[autos['price'].between(1000,450000)]
autos_valuate

dateCrawled                                          name  \
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
6       2016-04-01 20:48:51                   Peugeot_206_CC_110_Platinum   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  registration_year  \
1       privat   Angebot  18300     test       coupe               2011   
2       privat   Angebot   9800     test         suv               2004   
3       privat   Angebot   1500     test  kleinwagen               2001   
4       privat   Angebot   3600     test  kleinwagen               2008   
6       privat   Angebot   2200     test      cabrio               2004   
...        ...       ...    ...      ...         ...                ...   
371523  privat   Angebot   2200     test         NaN               2005   
371524  privat   Angebot   1199     test      cabrio               2000   
371525  privat   Angebot   9200     test         bus               1996   
371526  privat   Angebot   3400     test       kombi               2002   
371527  privat   Angebot  28990  control   limousine               2013   

          gearbox  powerPS        model  odometer_km  registration_month  \
1         manuell      190          NaN       125000                   5   
2       automatik      163        grand       125000                   8   
3         manuell       75         golf       150000                   6   
4         manuell       69        fabia        90000                   7   
6         manuell      109      2_reihe       150000                   8   
...           ...      ...          ...          ...                 ...   
371523        NaN        0          NaN        20000                   1   
371524  automatik      101       fortwo       125000                   3   
371525    manuell      102  transporter       150000                   3   
371526    manuell      100         golf       150000                   6   
371527    manuell      320      m_reihe        50000                   8   

       fuelType           brand unrepaired_damage           ad_created  \
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
6        benzin         peugeot              nein  2016-04-01 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
6                  0       67112  2016-04-05 18:18:39  
...              ...         ...                  ..